src/Umbraco.Core/Exceptions/ArgumentNullOrEmptyException.cs

```diff
@@ -1,30 +1,54 @@
 ﻿using System;
+using System.Runtime.Serialization;
 
 namespace Umbraco.Core.Exceptions
 {
     /// <summary>
-    /// The exception that is thrown when a null reference, or an empty argument,
-    /// is passed to a method that does not accept it as a valid argument.
+    /// The exception that is thrown when a null reference, or an empty argument, is passed to a method that does not accept it as a valid argument.
     /// </summary>
+    /// <seealso cref="System.ArgumentNullException" />
+    [Serializable]
     public class ArgumentNullOrEmptyException : ArgumentNullException
```

Comment:
I see this exception is used quite a lot, but it's semantically not correct. If this is thrown for empty values, the base class is `ArgumentNullException`, but the argument isn't `null`... And as the current value isn't available nor the message specific enough, debugging whether it was `null` or empty will be unnecessarily hard.

Yes, the following code is quite cumbersome to write, but it's correct and gives good debug information:
```c#
if (param == null) throw new ArgumentNullException(nameof(param));
if (string.IsNullOrWhiteSpace(param)) throw new ArgumentException("Value can't be empty or consist only of white-space characters.", nameof(param));
```

See also:
- https://docs.microsoft.com/en-us/dotnet/api/system.argumentexception (explains the differences between the exceptions)
- https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/code-contracts (might be a better way to do these validations)

Reply:
I wasn't aware of this exception or why it exists. Both Visual Studio and Resharper offer a very easy way to check strings for both null and empty by just pressing ctrl + enter. I personally don't see a reason for this exception but i don't see a reason to throw both an ArgumentNullException and an ArgumentException.

Reply:
Visual Studio and Resharper both generate this single check for null/empty strings:

Reply:
Well, the linked documentation from Microsoft clearly states:

> The primary derived classes of `ArgumentException` are `ArgumentNullException` and `ArgumentOutOfRangeException`. These derived classes should be used instead of `ArgumentException`, except in situations where neither of the derived classes is acceptable. For example, exceptions should be thrown by:
> 
> * `ArgumentNullException` whenever `null` is passed to a method that does not accept it as a valid argument.
> 
> * `ArgumentOutOfRangeException` when the value of an argument is outside the range of acceptable values; for example, when the value "46" is passed as the month argument during the creation of a `DateTime`.

This PR is all about best-practices, consistency and code-style, so I would recommend throwing the different exceptions for `null` and empty/white-space values. This also makes debugging a lot easier: just check what exception type was thrown to see what went wrong...

Reply:
Ok that's fine but we'll have to log a bug with MS themselves since they are generating code that doesn't adhere to their own standards ;) I think R# can do both exceptions but would need to check. Point is, that these null checks are generally performed by the tooling and that will continue to happen. There's are hundreds of these IsNullOrWhitespace checks with ArgumentException thrown because the tooling provides that. Perhaps the reason for this custom ArgumentNullOrEmptyException was to try to make this check clearer, i dunno.

Reply:
Visual Studio 2017 (15.9.15) generates the correct checks, so it's probably something in ReSharper or customized settings:
![ArgumentNullException](https://user-images.githubusercontent.com/1051287/63718316-9685d880-c84a-11e9-9454-3aadb693f36d.png)

![ArgumentException](https://user-images.githubusercontent.com/1051287/63718502-072cf500-c84b-11e9-917c-db9d22c13b68.png)